File: lib/src/transport.dart
import 'package:http/http.dart' as http;

import 'adapters.dart';

/// Default [Transport] using `package:http` (D18). Injectable for tests.
class HttpTransport implements Transport {
  HttpTransport({
    http.Client? client,
    this.timeout = const Duration(seconds: 30),
  })  : _client = client ?? http.Client(),
        _ownsClient = client == null;

  final http.Client _client;
  final Duration timeout;
  final bool _ownsClient;

  @override
  Future<TransportResponse> send(TransportRequest request) async {
    final req = http.Request(request.method, request.url);
    req.followRedirects = true;
    req.persistentConnection = true;
    request.headers.forEach((key, value) {
      req.headers[key] = value;
    });
    if (request.body != null) {
      req.bodyBytes = request.body!;
    }
    final streamed = await _client.send(req).timeout(timeout);
    final bytes = await streamed.stream.toBytes().timeout(timeout);
    return TransportResponse(
      statusCode: streamed.statusCode,
      headers: Map<String, String>.from(streamed.headers),
      body: bytes,
    );
  }

  void close() {
    if (_ownsClient) {
      _client.close();
    }
  }
}

File: src/transport.ts
export interface TransportRequest {
  method: string;
  url: string;
  headers: Record<string, string>;
  body?: Uint8Array | null;
  signal?: AbortSignal;
}

export interface TransportResponse {
  status: number;
  headers: Record<string, string>;
  body: Uint8Array;
}

/** HTTP GET/HEAD/POST with Range and preserved query strings. */
export interface Transport {
  request(req: TransportRequest): Promise<TransportResponse>;
}

function toHeaderRecord(headers: Headers): Record<string, string> {
  const out: Record<string, string> = {};
  headers.forEach((value, key) => {
    out[key.toLowerCase()] = value;
  });
  return out;
}

/** Node 20+ / Electron app-level `globalThis.fetch`. Do not use axios or node-fetch. */
export class FetchTransport implements Transport {
  constructor(private readonly fetchImpl: typeof fetch = globalThis.fetch.bind(globalThis)) {}

  async request(req: TransportRequest): Promise<TransportResponse> {
    const headers = new Headers(req.headers);
    const init: RequestInit = {
      method: req.method,
      headers,
      signal: req.signal,
    };
    if (req.body && req.method !== "GET" && req.method !== "HEAD") {
      init.body = req.body;
    }
    const res = await this.fetchImpl(req.url, init);
    const body = new Uint8Array(await res.arrayBuffer());
    return { status: res.status, headers: toHeaderRecord(res.headers), body };
  }
}
